firmware/controllers/actuators/electronic_throttle.cpp

```diff
@@ -269,6 +274,29 @@ static percent_t currentEtbDuty;
 
 		etb1.dcMotor.Set(ETB_PERCENT_TO_DUTY(currentEtbDuty));
 
+	   	float etbError = absF(targetPosition - actualThrottlePosition); // Used for I-Term limiter and Error-detection
+
+
+        // ETB Deviation monitoring
+        if(etbError < 3 ){
+          start_time = chVTGetSystemTimeX(); //Updates timestamp as long as ETB-Error stays within 3% deviation
+        }
+      else {
+        }        // ETB Deviation outside limit detected, set timestamp and monitor it for two seconds
+        if(chVTIsSystemTimeWithinX(start_time, start_time + TIME_MS2I(2000))) {
+        	  //Less than two seconds, system OK: do nothing
+            }
+       else {
+         if(rpm > 600){  //We dont want or need any of this to execute if engine are not running
+                         // It would make a huge issue while tuning ETB
+        	 etbErrorCondition = true;
+        	   warning(OBD_Throttle_Actuator_Control_Range_Performance_Bank_1, "ETB Deviation Error"); //Set Error-code
+        	   engineConfiguration->rpmHardLimit = 2000;  //Set RPM Limit to 2000 to not kill engine if ETB opening are too big
```

Comment:
Logic flow should never write into settings like that "engineConfiguration->rpmHardLimit"

if you need to use a setting and some logic, you need engine->rpmHardLimit variable for CURRENT rpmHardLimit , or, better, "if LIMPING_MODE then ONE RPM LIMIT LOGIC else NORMAL SETTING"

Reply:
This is the part I haven't found the exact way how to do it, so a little help would be nice.

If the flow comes to this condition, I would like it to stay like this until a reset.

As is now, it passes on every time etberror are within the allowed area, and triggers again.

I tried with a while-loop, but that shuts the whole firmware.

I only want the rpm-limit, error message and duty = 0 to stay.

Reply:
I was thinking having only etbErrorCondition == true; at the end.

Then have another function to define if etbError = true.

Could that be a idea?

Reply:
in order for unrelated files to share this variable you should define it next to hwTriggerInputEnabled on Engine or even better on EngineState next to engineNoiseHipLevel

I would say in this case you have

etbDeviationError = false
on start-up

and if you ever get in deviation error state you say etbDeviationError  = true


in the code which reads rpmHardLimit you would have to read etbDeviationError  as well

Reply:
For now maybe don't worry about rev limiting - we need a more cohesive "limp home" strategy for broader categories of errors.  Implementing every possible condition that should trigger limp home is going to cause a spaghetti mess.